src/primitives/compression.cpp

```diff
@@ -0,0 +1,283 @@
+          CScript redeemScript = CScript(vec.begin(), vec.end());
+          if (redeemScript.IsPayToWitnessKeyHash()) {
+              scriptSig = CScript(txin.scriptWitness.stack.at(0).begin(), txin.scriptWitness.stack.at(0).end());
+          }
+        }
+
+        if (scriptSig.size()) {
+            opcodetype opcodeRet;
+            CScriptBase::const_iterator pc = scriptSig.begin();
+            scriptSig.GetOp(pc, opcodeRet, m_signature);
+            uint8_t hashType = m_signature.at(m_signature.size()-1);
+            int length = m_signature.size()-1;
+            secp256k1_ecdsa_signature sig;
+            if (secp256k1_ecdsa_signature_parse_der(secp256k1_context_static, &sig, &m_signature[0], length)) {
+                if (secp256k1_ecdsa_signature_serialize_compact(secp256k1_context_static, &m_signature[0], &sig)) {
+                    m_signature.resize(64);
+                    m_compressedSignature = true;
+                    m_hashType = hashType;
+                    m_isHashStandard = m_hashType == 0x01;
+                    if (m_isHashStandard) m_hashType = 0;
+                }
+            }
+        }
+
+        if (!m_compressedSignature) {
+            CDataStream stream(SER_DISK, 0);
+            stream << VARINT(txin.scriptSig.size());
+            if (txin.scriptSig.size())
+                stream.write(MakeByteSpan(txin.scriptSig));
+            stream << VARINT(txin.scriptWitness.stack.size());
+            for (size_t index = 0; index < txin.scriptWitness.stack.size(); index++) {
+                stream << VARINT(txin.scriptWitness.stack.at(index).size());
+                stream.write(MakeByteSpan(txin.scriptWitness.stack.at(index)));
+            }
+            m_signature.resize(stream.size());
+            stream.read(MakeWritableByteSpan(m_signature));
+        }
+    }
+    m_nSequence = txin.nSequence;
+}
+
+std::string CCompressedTxIn::ToString() const
+{
+    return strprintf("CCompressedTxIn(prevout=%s, signature=%s, compressed=%b, hashType=%u, isHashStandard=%b, nSequence=%u)",
+    m_prevout.ToString(),
+    HexStr(m_signature),
+    m_compressedSignature,
+    m_hashType,
+    m_isHashStandard,
+    m_nSequence);
+}
+
+CCompressedTxOut::CCompressedTxOut(const CTxOut& txout) {
+    m_nValue = txout.nValue;
+    std::vector<std::vector<unsigned char>> solutions;
+    m_scriptType = Solver(txout.scriptPubKey, solutions);
+    switch (m_scriptType) {
+        case TxoutType::PUBKEY:
+        case TxoutType::PUBKEYHASH:
+        case TxoutType::SCRIPTHASH:
+        case TxoutType::WITNESS_V0_KEYHASH:
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+        case TxoutType::WITNESS_V1_TAPROOT:
+            m_scriptPubKey = solutions.at(0);
+            break;
+        default:
+            m_scriptType = TxoutType::NONSTANDARD;
+            m_scriptPubKey.resize(txout.scriptPubKey.size());
+            copy(txout.scriptPubKey.begin(), txout.scriptPubKey.end(), m_scriptPubKey.begin());
```

Comment:
It complains about an overloaded '='. Its because the CScript is a prevector which is another custom vector type in bitcoin core. I can use .data() to get ptr to a vector of the elements, but I still have to resize and set the ptr. Thanks- Tom.

Reply:
Hmm, ok. There are many instances in the codebase of scripts being directly assigned like this. I'm not sure what's wrong here.